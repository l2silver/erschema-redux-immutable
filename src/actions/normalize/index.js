// @flow
import normalize from 'erschema';
import {batchActions} from 'redux-batched-actions';
import {retypeAction} from 'redux-retype-actions';
import {get} from 'lodash';
import {generateActionName} from 'resource-action-types';
import entityActions from '../entities'
import relationshipActions from '../relationships'


import type {$schema} from 'erschema/types';

export function normalizeActions(entity: Object, name: string, schema: $schema, firstSchema?: $schema, options?: Object = {}){
  const {entities: allEntities, relationships: allRelationships} = normalize(entity, name, schema, firstSchema)
  const indexEntities = Object.keys(allEntities).reduce((finalResult, entityName)=>{
    const entities = allEntities[entityName]
    const entitiesForIndex = Object.keys(entities).reduce((finalResult, entityId)=>{
      finalResult.push(entities[entityId])
      return finalResult
    }, [])
    finalResult.push(entityActions.index(entityName, entitiesForIndex))
    return finalResult
  }, [])
  const indexRelationships = Object.keys(allRelationships).reduce((allRelationshipActions, entityName)=>{
    const relationships = allRelationships[entityName]
    const relationshipsForIndex = Object.keys(relationships).reduce((finalResult, relationshipName)=>{
      const relationship = relationships[relationshipName]
      const idValuePairs = Object.keys(relationship).reduce((finalResult, entityId)=>{
        finalResult.push({id: entityId, value: relationship[entityId]})
        return finalResult
      }, [])
      
      const relationshipActionName = get(options, `relationships.${entityName === 'pages' ? idValuePairs[0].id : entityName}.${relationshipName}.concat`) ? 'concatRelationship' : 'index'
      finalResult.push(relationshipActions[relationshipActionName](entityName, {name: relationshipName, idValuePairs}))
      return finalResult
    }, allRelationshipActions)
    return allRelationshipActions
  }, [])
  return {
    indexEntities,
    indexRelationships
  }
}

export function indexNormalizeActions(entities: Object[], name: string, schema: $schema, firstSchema?: $schema, options?: {}){
  return entities.map((entity)=>{
    return normalizeActions(entity, name, schema, firstSchema, options)
  })
}