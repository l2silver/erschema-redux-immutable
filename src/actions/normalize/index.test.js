// @flow

import normalize from './'

describe('normalize', ()=>{
  it('return an object with indexRelations and indexEntities', ()=>{
    const schema = {
      users: {

      }
    }
    const user = {
      id: 1,
      name: 'Jimmy'
    }
    expect(normalize(user, 'users', schema)).toMatchSnapshot()
  })
})