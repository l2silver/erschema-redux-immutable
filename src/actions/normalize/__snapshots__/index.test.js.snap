// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize return an object with indexRelations and indexEntities 1`] = `
Object {
  "indexEntities": Array [
    Object {
      "error": undefined,
      "payload": Object {
        "entities": Array [
          Object {
            "id": 1,
            "name": "Jimmy",
          },
        ],
      },
      "type": "INDEX_USER",
    },
  ],
  "indexRelationships": Array [],
}
`;
